modules/dataquery/jsx/hooks/usewidgets.tsx

```diff
@@ -0,0 +1,103 @@
+import {useState, useEffect} from 'react';
+import {Tabs, TabSteps} from '../nextsteps';
+
+/**
+ * Converts a string name of a callback function to the callback
+ * itself.
+ *
+ * @param {string} name - The name of the callback function
+ * @returns {Function | null} - The callback if available
+ */
+function getCallback(name: string): (() => void) | null {
+    const namepieces = name.split('.').reverse();
+    let level: any = window;
+    for (let name = namepieces.pop(); name; name = namepieces.pop()) {
+         if (level[name]) {
+             level = level[name];
+         } else {
+             // We have not reached the end, but the
+             // name was not defined, return null
+             return null;
+         }
+    }
+    return level;
+}
+
+/**
+ * React hook to load a list of valid visits from the server
+ * and manage which should be selected by default
+ *
+ * @returns {TabSteps} - list of default and all visits
+ */
+function useWidgets(): TabSteps {
+    const [steps, setSteps] = useState<TabSteps>({});
+    useEffect(() => {
+          fetch('/dataquery/widgets', {credentials: 'same-origin'})
+          .then((resp) => {
+                  if (!resp.ok) {
+                      throw new Error('Invalid response');
+                  }
+                  return resp.json();
+          }).then((result) => {
+              const steps: TabSteps = {
+                  [Tabs.Info]: [],
+                  [Tabs.Fields]: [],
+                  [Tabs.Filters]: [],
+                  [Tabs.Data]: [],
+              };
+              for (const widget of result) {
+                  const func = getCallback(widget.callback);
+                  if (func === null) {
+                      console.error(
+                          'Unknown callback function',
+                          widget.callback
+                      );
+                      continue;
+                  }
+
+                  switch (widget.tab) {
+                  case 'info':
+                      if (steps[Tabs.Info]) {
+                          steps[Tabs.Info].push({
+                             label: widget.label,
+                             action: func,
+                          });
+                      }
+                      break;
```

Comment:
The pattern of this `switch` should maybe be factorized ?

Reply:
What do you mean? The switch is delegating which bucket to put the item in. It's one if statement per case.